Library/Homebrew/migrator.rb

```diff
@@ -0,0 +1,327 @@
+require "formula"
+require "keg"
+require "tab"
+
+class Migrator
+  class MigratorNoOldnameError < RuntimeError
+    def initialize(formula)
+      super "#{formula.name} doesn't replace any formula."
+    end
+  end
+
+  class MigratorNoOldpathError < RuntimeError
+    def initialize(formula)
+      super "#{HOMEBREW_CELLAR/formula.oldname} doesn't exist."
+    end
+  end
+
+  class MigratorDifferentTapsError < RuntimeError
+    def initialize(formula, tap)
+      if tap.nil?
+        super <<-EOS.undent
+        #{formula.name} from #{formula.tap} is given, but old name #{formula.oldname} wasn't installed from taps or core formulae
+        EOS
+      else
+        user, repo = tap.split("/")
+        repo.sub!("homebrew-", "")
+        name = "fully-qualified #{user}/#{repo}/#{formula.oldname}"
+        name = formula.oldname if tap == "Homebrew/homebrew"
+        super <<-EOS.undent
+        #{formula.name} from #{formula.tap} is given, but old name #{formula.oldname} was installed from #{tap}
+
+        Please try to use #{name} to refer the formula
+        EOS
+      end
+    end
+  end
+
+  attr_reader :formula
+  attr_reader :oldname, :oldpath, :old_pin_record, :old_opt_record
+  attr_reader :old_linked_keg_record, :oldkeg, :old_tabs, :old_tap
+  attr_reader :newname, :newpath, :new_pin_record
+  attr_reader :old_pin_link_record
+
+  def initialize(formula)
+    raise MigratorNoOldnameError.new(formula) unless formula.oldname
+
+    @formula = formula
+    @oldpath = HOMEBREW_CELLAR/formula.oldname
+    raise MigratorNoOldpathError.new(formula) unless oldpath.exist?
+
+    @old_tabs = oldpath.subdirs.each.map { |d| Tab.for_keg(Keg.new(d)) }
+    @old_tap = old_tabs.first.tap
+    raise MigratorDifferentTapsError.new(formula, old_tap) unless from_same_taps?
+
+    @newpath = HOMEBREW_CELLAR/formula.name
+    @oldname = formula.oldname
+    @newname = formula.name
+
+    if @oldkeg = get_linked_oldkeg
+      @old_linked_keg_record = oldkeg.linked_keg_record if oldkeg.linked?
+      @old_opt_record = oldkeg.opt_record if oldkeg.optlinked?
+    end
+
+    @old_pin_record = HOMEBREW_LIBRARY/"PinnedKegs"/oldname
+    @new_pin_record = HOMEBREW_LIBRARY/"PinnedKegs"/newname
+    @pinned = old_pin_record.symlink?
+    @old_pin_link_record = old_pin_record.readlink if @pinned
+  end
+
+  # Check if installed keg is from core formulae and newname is from core.
+  def both_from_core?
+    new_core = formula.core_formula?
+    old_core = (old_tap.nil? || old_tap == "Homebrew/homebrew" \
+      || old_tap == "mxcl/master")
+    # Formula can be installed from path and we have to check wheter this path
+    # is the same as for core formula.
+    if (path = old_tabs.first.source["path"]) \
+        && path != Formulary.core_path(formula.oldname).to_s
+      old_core = false
+    end
```

Comment:
I think you can remove this condition block. Because `Formula#tap` is resolved based on `Formula#path`, it's impossible to have tap is unknown while `Tab#source["path"]` is pointed to a core formula.